src/Core/IdentityServer/BaseRequestValidator.cs

```diff
@@ -259,11 +274,63 @@ protected async Task BuildErrorResultAsync(bool twoFactorRequest, T context, Use
             return new Tuple<bool, Organization>(individualRequired || firstEnabledOrg != null, firstEnabledOrg);
         }
 
+        private async Task<bool> RequiresSsoAuthAsync(User user, string grantType)
+        {
+            if (grantType == "authorization_code")
+            {
+                return true; // Already using SSO to authorize, finish successfully
+            }
+            
+            // Is user apart of any orgs? Use cache for initial checks.
+            var orgs = (await _currentContext.OrganizationMembershipAsync(_organizationUserRepository, user.Id))
+                .ToList();
+            if (orgs.Any())
+            {
+                // Get all org abilities
+                var orgAbilities = await _applicationCacheService.GetOrganizationAbilitiesAsync();
+                // Parse all user orgs that are enabled and have the ability to use sso
+                var ssoOrgs = orgs.Where(o => OrgCanUseSso(orgAbilities, o.Id));
+                if (ssoOrgs.Any())
+                {
+                    // Parse users orgs and determine if require sso policy is enabled
+                    var userOrgs = await _organizationRepository.GetManyByUserIdAsync(user.Id);
+                    var requiresSso = false;
+                    foreach (var org in userOrgs)
+                    {
+                        if (!(org.Enabled && org.UseSso))
+                        {
+                            continue;
+                        }
+                        
+                        var orgPolicy = await _policyRepository.GetByOrganizationIdTypeAsync(org.Id, PolicyType.RequireSso);
+                        if (orgPolicy != null && orgPolicy.Enabled)
+                        {
+                            requiresSso = true;
+                            break;
+                        }
+                    }
+
+                    if (requiresSso)
```

Comment:
this `if` block can be removed altogether

Reply:
Will update

Reply:
Updated.